pbft/pbft.cpp

```diff
@@ -994,6 +1007,533 @@ pbft::get_high_water_mark()
+        {
+            return false;
+        }
+
+        viewchange_senders.insert(original_msg.sender());
+        std::set<std::pair<uint64_t, std::string>> sequence_request_pairs;
+
+        if (!this->get_sequences_and_request_hashes_from_proofs(viewchange_msg, sequence_request_pairs))
+        {
+            return false;
+        }
+
+        // each *pair* must match a pre-prepare in the the newview message...
+        std::vector<pbft_msg> matched_pre_prepares;
+        for (const auto sequence_requesthash : sequence_request_pairs)
+        {
+            uint64_t sequence{sequence_requesthash.first};
+            std::string request_hash{sequence_requesthash.second};
+
+            for (int j{0}; j <= msg.pre_prepare_messages_size(); ++j)
+            {
+                const bzn_envelope& pre_prepare_envelope = msg.pre_prepare_messages(j);
+                // TODO: do we need to cryptographically validate pre_prepare_envelope?
+                // TODO: do we need to check the sender?
+
+                pbft_msg pre_prepare;
+                if (!pre_prepare.ParseFromString(pre_prepare_envelope.pbft()))
+                {
+                    LOG (error) << "is_valid_newview_message - pre_prepare not valid";
+                    return false;
+                }
+
+                // did we find the sequence?
+                if (sequence == pre_prepare.sequence()
+                    && request_hash == pre_prepare.request_hash())
+                {
+                    matched_pre_prepares.emplace_back(pre_prepare);
+                    break;
+                }
+            }
+        }
+
+        if (matched_pre_prepares.size() != size_t(msg.pre_prepare_messages_size()))
+        {
+            LOG (error) << "is_valid_newview_message - unexpected sequence/request hash count";
+            return false;
+        }
+
+        if (sequence_request_pairs.size() != size_t(msg.pre_prepare_messages_size()))
+        {
+            LOG (error) << "is_valid_newview_message - pre prepare without matching hash";
+            return false;
+        }
+    }
+
+    if (viewchange_senders.size() != size_t(msg.viewchange_messages_size()))
+    {
+        LOG (error) << "is_valid_newview_message - unexpected viewchange message count";
+        return false;
+    }
+
+    return true;
+}
+
+void
+pbft::fill_in_missing_pre_prepares(std::map<uint64_t, bzn_envelope> &pre_prepares)
+{
+    uint64_t last_sequence_number = pre_prepares.empty() ? 0 : (pre_prepares.end()--)->first;
+    for (uint64_t i = this->latest_stable_checkpoint().first + 1; i < last_sequence_number; ++i)
+    {
+        //  -- create a new preprepare for a no-op operation using this sequence number
+        if (pre_prepares.find(i) == pre_prepares.end())
+        {
+            auto msg = new database_msg;
+            msg->set_allocated_nullmsg(new database_nullmsg);
```

Comment:
this is a database_msg, not a preprepare

Reply:
Paul helped me rewrite this section.